OpenRA.Game/Graphics/WorldRenderer.cs

```diff
@@ -107,78 +109,127 @@ public void ReplacePalette(string name, IPalette pal)
+					renderablesBuffer.Add(renderable);
+
+			foreach (var renderable in World.WorldActor.Render(this))
+				renderablesBuffer.Add(renderable);
+
 			if (World.RenderPlayer != null)
-				actors = actors.Append(World.RenderPlayer.PlayerActor);
+				foreach (var renderable in World.RenderPlayer.PlayerActor.Render(this))
+					renderablesBuffer.Add(renderable);
 
-			var worldRenderables = actors.SelectMany(a => a.Render(this));
 			if (World.OrderGenerator != null)
-				worldRenderables = worldRenderables.Concat(World.OrderGenerator.Render(this, World));
+				foreach (var renderable in World.OrderGenerator.Render(this, World))
+					renderablesBuffer.Add(renderable);
 
 			// Unpartitioned effects
-			worldRenderables = worldRenderables.Concat(World.UnpartitionedEffects.SelectMany(e => e.Render(this)));
+			foreach (var e in World.UnpartitionedEffects)
+				foreach (var renderable in e.Render(this))
+					renderablesBuffer.Add(renderable);
 
 			// Partitioned, currently on-screen effects
-			var effectRenderables = World.ScreenMap.RenderableEffectsInBox(Viewport.TopLeft, Viewport.BottomRight);
-			worldRenderables = worldRenderables.Concat(effectRenderables.SelectMany(e => e.Render(this)));
+			foreach (var e in World.ScreenMap.RenderableEffectsInBox(Viewport.TopLeft, Viewport.BottomRight))
+				foreach (var renderable in e.Render(this))
+					renderablesBuffer.Add(renderable);
 
-			worldRenderables = worldRenderables.OrderBy(RenderableScreenZPositionComparisonKey);
+			foreach (var renderable in renderablesBuffer.OrderBy(RenderableScreenZPositionComparisonKey))
+				preparedRenderables.Add(renderable.PrepareRender(this));
 
-			return worldRenderables.Select(r => r.PrepareRender(this));
+			renderablesBuffer.Clear();
 		}
 
-		IEnumerable<IFinalizedRenderable> GenerateOverlayRenderables()
+		void GenerateOverlayRenderables()
 		{
-			var actors = World.ActorsWithTrait<IRenderAboveShroud>()
-				.Where(a => a.Actor.IsInWorld && !a.Actor.Disposed && (!a.Trait.SpatiallyPartitionable || onScreenActors.Contains(a.Actor)))
-					.SelectMany(a => a.Trait.RenderAboveShroud(a.Actor, this));
+			// actors
+			foreach (var a in World.ActorsWithTrait<IRenderAboveShroud>())
+			{
+				if ((!a.Actor.IsInWorld || a.Actor.Disposed) && (a.Trait.SpatiallyPartitionable && onScreenActors.Contains(a.Actor)))
+					continue;
 
-			var selected = World.Selection.Actors.Where(a => a.IsInWorld && !a.Disposed)
-				.SelectMany(a => a.TraitsImplementing<IRenderAboveShroudWhenSelected>()
-					.Where(t => !t.SpatiallyPartitionable || onScreenActors.Contains(a))
-					.SelectMany(t => t.RenderAboveShroud(a, this)));
+				foreach (var renderable in a.Trait.RenderAboveShroud(a.Actor, this))
+					preparedOverlayRenderables.Add(renderable.PrepareRender(this));
+			}
 
-			var effects = World.Effects.Select(e => e as IEffectAboveShroud)
-				.Where(e => e != null)
-				.SelectMany(e => e.RenderAboveShroud(this));
+			// selected
```

Comment:
IMO this comment (and maybe the actors one above) don't add much so could just be removed.